File: translated/f-to-c/DRB062-matrixvector2-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Matrix-vector multiplication: inner level parallelization. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

void foo()
{
    int i, j, N;
    float sum;
    float** a;
    float* v;
    float* v_out;

    N = 1000;

    // Allocate arrays
    a = (float**)malloc(N * sizeof(float*));
    for (i = 0; i < N; i++) {
        a[i] = (float*)malloc(N * sizeof(float));
    }
    v = (float*)malloc(N * sizeof(float));
    v_out = (float*)malloc(N * sizeof(float));

    for (i = 0; i < N; i++) {
        sum = 0.0;
        #pragma omp parallel for reduction(+:sum)
        for (j = 0; j < N; j++) {
            sum = sum + a[i][j] * v[j];
            printf("%f\n", sum);
        }
        v_out[i] = sum;
    }

    // Free arrays
    for (i = 0; i < N; i++) {
        free(a[i]);
    }
    free(a);
    free(v);
    free(v_out);
}

int main()
{
    foo();
    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB062-matrixvector2-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Matrix-vector multiplication: inner level parallelization. No data race pairs.

package main

import (
	"fmt"
	"runtime"
	"sync"
)

func foo() {
	var i, N int
	var sum float64
	var a [][]float64
	var v, v_out []float64

	N = 1000
	a = make([][]float64, N)
	for i := range a {
		a[i] = make([]float64, N)
	}
	v = make([]float64, N)
	v_out = make([]float64, N)

	for i = 1; i <= N; i++ {
		sum = 0.0
		//$omp parallel do reduction(+:sum)
		var wg sync.WaitGroup
		var sumMutex sync.Mutex
		numCPU := runtime.NumCPU()
		chunkSize := N / numCPU
		if chunkSize < 1 {
			chunkSize = 1
		}

		for start := 1; start <= N; start += chunkSize {
			end := start + chunkSize - 1
			if end > N {
				end = N
			}
			wg.Add(1)
			go func(i, start, end int) {
				defer wg.Done()
				localSum := 0.0
				for j := start; j <= end; j++ {
					localSum = localSum + a[i-1][j-1]*v[j-1]
					fmt.Println(localSum)
				}
				sumMutex.Lock()
				sum += localSum
				sumMutex.Unlock()
			}(i, start, end)
		}
		wg.Wait()
		//$omp end parallel do
		v_out[i-1] = sum
	}
}

func main() {
	foo()
}
